File: hkoweather_bot/hko.py
from __future__ import annotations

import logging
from dataclasses import dataclass
from datetime import datetime
from typing import Any

import requests

LOGGER = logging.getLogger(__name__)


class HkoClientError(RuntimeError):
    """Raised when Hong Kong Observatory data cannot be fetched."""


@dataclass(frozen=True)
class CurrentWeather:
    temperature: str
    humidity: str
    uv_index: str
    rainfall: str
    update_time: str
    warning_message: str


@dataclass(frozen=True)
class WarningSnapshot:
    signature: str
    message: str
    has_notification: bool


class HkoClient:
    BASE_URL = "https://data.weather.gov.hk/weatherAPI/opendata"

    def __init__(self, language: str = "tc", timeout_seconds: int = 15) -> None:
        self.language = language
        self.timeout_seconds = timeout_seconds

    def _get_json(self, endpoint: str, params: dict[str, str] | None = None) -> dict[str, Any]:
        merged_params = {"lang": self.language}
        if params:
            merged_params.update(params)

        try:
            response = requests.get(
                f"{self.BASE_URL}/{endpoint}",
                params=merged_params,
                timeout=self.timeout_seconds,
            )
            response.raise_for_status()
            return response.json()
        except requests.RequestException as exc:
            raise HkoClientError(f"Failed to fetch HKO endpoint {endpoint}: {exc}") from exc
        except ValueError as exc:
            raise HkoClientError(f"HKO endpoint {endpoint} returned invalid JSON") from exc

    def get_current_weather(self) -> CurrentWeather:
        data = self._get_json("weather.php", {"dataType": "rhrread"})

        temperature = self._format_temperature(data.get("temperature"))
        humidity = self._format_humidity(data.get("humidity"))
        uv_index = self._format_uv_index(data.get("uvindex"))
        rainfall = self._format_rainfall(data.get("rainfall"))
        update_time = self._format_update_time(data.get("updateTime"))
        warning_message = self._format_warning_message(data)

        return CurrentWeather(
            temperature=temperature,
            humidity=humidity,
            uv_index=uv_index,
            rainfall=rainfall,
            update_time=update_time,
            warning_message=warning_message,
        )

    def get_warning_snapshot(self) -> WarningSnapshot:
        active_warnings = self._get_json("weather.php", {"dataType": "warnsum"})
        warning_info = self._get_json("weather.php", {"dataType": "warningInfo"})

        lines: list[str] = []
        signature_parts: list[str] = []

        if active_warnings:
            lines.append("現正生效天氣警告：")
            for code in sorted(active_warnings):
                item = active_warnings.get(code)
                if not isinstance(item, dict):
                    continue
                name = str(item.get("name") or code)
                action_code = str(item.get("actionCode") or "")
                issue_time = str(item.get("issueTime") or "")
                expire_time = str(item.get("expireTime") or "")
                lines.append(f"- {name}")
                if issue_time:
                    lines.append(f"  發出時間：{self._format_datetime(issue_time)}")
                if expire_time:
                    lines.append(f"  到期時間：{self._format_datetime(expire_time)}")
                signature_parts.append(f"{code}:{name}:{action_code}:{issue_time}:{expire_time}")

        details = warning_info.get("details")
        if isinstance(details, list):
            for detail in details:
                if not isinstance(detail, dict):
                    continue
                contents = detail.get("contents")
                if not isinstance(contents, list) or not contents:
                    continue
                warning_statement_code = str(detail.get("warningStatementCode") or "")
                update_time = str(detail.get("updateTime") or "")
                text = "\n".join(str(content).strip() for content in contents if str(content).strip())
                if not text:
                    continue
                if lines:
                    lines.append("")
                lines.append(text)
                signature_parts.append(f"{warning_statement_code}:{update_time}:{text}")

        signature = "|".join(signature_parts)
        return WarningSnapshot(
            signature=signature,
            message="\n".join(lines).strip(),
            has_notification=bool(signature_parts),
        )

    @staticmethod
    def _format_temperature(raw: Any) -> str:
        records = raw.get("data") if isinstance(raw, dict) else None
        if not isinstance(records, list):
            return "未能取得"

        hko_record = _find_record(records, place="香港天文台")
        if hko_record is None and records:
            hko_record = records[0] if isinstance(records[0], dict) else None
        if not hko_record:
            return "未能取得"

        value = hko_record.get("value")
        unit = raw.get("unit", "C") if isinstance(raw, dict) else "C"
        place = hko_record.get("place", "香港")
        return f"{place} {value}°{unit}" if value is not None else "未能取得"

    @staticmethod
    def _format_humidity(raw: Any) -> str:
        records = raw.get("data") if isinstance(raw, dict) else None
        if not isinstance(records, list) or not records:
            return "未能取得"
        record = records[0] if isinstance(records[0], dict) else {}
        value = record.get("value")
        unit = raw.get("unit", "%") if isinstance(raw, dict) else "%"
        return f"{value}{unit}" if value is not None else "未能取得"

    @staticmethod
    def _format_uv_index(raw: Any) -> str:
        records = raw.get("data") if isinstance(raw, dict) else None
        if not isinstance(records, list) or not records:
            return "未能取得"
        record = records[0] if isinstance(records[0], dict) else {}
        value = record.get("value")
        desc = record.get("desc")
        if value is None:
            return "未能取得"
        return f"{value}（{desc}）" if desc else str(value)

    @staticmethod
    def _format_rainfall(raw: Any) -> str:
        records = raw.get("data") if isinstance(raw, dict) else None
        if not isinstance(records, list):
            return "未能取得"

        rainy_records = [
            record
            for record in records
            if isinstance(record, dict) and record.get("max", 0) not in (0, "0", None)
        ]
        if not rainy_records:
            return "過去一小時大部分地區沒有錄得雨量"

        unit = raw.get("unit", "mm") if isinstance(raw, dict) else "mm"
        formatted = []
        for record in rainy_records[:5]:
            place = record.get("place", "未知地區")
            minimum = record.get("min")
            maximum = record.get("max")
            if minimum == maximum or minimum in (None, ""):
                formatted.append(f"{place} {maximum}{unit}")
            else:
                formatted.append(f"{place} {minimum}-{maximum}{unit}")
        return "；".join(formatted)

    @staticmethod
    def _format_warning_message(data: dict[str, Any]) -> str:
        warning_message = data.get("warningMessage")
        if isinstance(warning_message, list) and warning_message:
            return "\n".join(str(item).strip() for item in warning_message if str(item).strip())
        return "沒有特別天氣提示"

    @staticmethod
    def _format_update_time(raw: Any) -> str:
        if not raw:
            return "未知"
        return HkoClient._format_datetime(str(raw))

    @staticmethod
    def _format_datetime(raw: str) -> str:
        try:
            return datetime.fromisoformat(raw.replace("Z", "+00:00")).strftime("%Y-%m-%d %H:%M")
        except ValueError:
            LOGGER.debug("Unable to parse datetime from HKO payload: %s", raw)
            return raw


def _find_record(records: list[Any], **criteria: str) -> dict[str, Any] | None:
    for record in records:
        if not isinstance(record, dict):
            continue
        if all(record.get(key) == value for key, value in criteria.items()):
            return record
    return None

File: src/core/hko.ts
import type {
  CurrentWeather,
  DailyDetailedContext,
  DailyWeatherContext,
  DayForecast,
  DetailedCurrentWeather,
  LocalForecast,
  NineDayForecastOverview,
  SpecialWeatherTips,
  StationReading,
  WarningSnapshot,
} from "./types";

const HKO_BASE_URL = "https://data.weather.gov.hk/weatherAPI/opendata/weather.php";

export async function getCurrentWeather(): Promise<CurrentWeather> {
  const data = await hkoJson<Record<string, unknown>>("rhrread");

  return {
    temperature: formatTemperature(data.temperature),
    humidity: formatHumidity(data.humidity),
    uvIndex: formatUvIndex(data.uvindex),
    rainfall: formatRainfall(data.rainfall),
    updateTime: formatUpdateTime(data.updateTime),
    warningMessage: formatWeatherWarningMessage(data.warningMessage),
  };
}

export async function getWarningSnapshot(): Promise<WarningSnapshot> {
  const activeWarnings = await hkoJson<Record<string, unknown>>("warnsum");
  const warningInfo = await hkoJson<Record<string, unknown>>("warningInfo");
  const lines: string[] = [];
  const signatureParts: string[] = [];

  if (Object.keys(activeWarnings).length > 0) {
    lines.push("現正生效天氣警告：");
    for (const code of Object.keys(activeWarnings).sort()) {
      const item = asRecord(activeWarnings[code]);
      if (!item) {
        continue;
      }

      const name = String(item.name || code);
      const actionCode = String(item.actionCode || "");
      const issueTime = String(item.issueTime || "");
      const expireTime = String(item.expireTime || "");

      lines.push(`- ${name}`);
      if (issueTime) {
        lines.push(`  發出時間：${formatDateTime(issueTime)}`);
      }
      if (expireTime) {
        lines.push(`  到期時間：${formatDateTime(expireTime)}`);
      }
      signatureParts.push(`${code}:${name}:${actionCode}:${issueTime}:${expireTime}`);
    }
  }

  const details = Array.isArray(warningInfo.details) ? warningInfo.details : [];
  for (const rawDetail of details) {
    const detail = asRecord(rawDetail);
    if (!detail || !Array.isArray(detail.contents)) {
      continue;
    }

    const text = detail.contents
      .map((content) => String(content).trim())
      .filter(Boolean)
      .join("\n");
    if (!text) {
      continue;
    }

    if (lines.length > 0) {
      lines.push("");
    }
    lines.push(text);
    signatureParts.push(
      `${String(detail.warningStatementCode || "")}:${String(detail.updateTime || "")}:${text}`,
    );
  }

  return {
    signature: signatureParts.join("|"),
    message: lines.join("\n").trim(),
    hasNotification: signatureParts.length > 0,
  };
}

export async function getLocalForecast(): Promise<LocalForecast> {
  const data = await hkoJson<Record<string, unknown>>("flw");

  return {
    generalSituation: String(data.generalSituation || "").trim(),
    tcInfo: String(data.tcInfo || "").trim(),
    fireDangerWarning: String(data.fireDangerWarning || "").trim(),
    forecastPeriod: String(data.forecastPeriod || "").trim(),
    forecastDesc: String(data.forecastDesc || "").trim(),
    outlook: String(data.outlook || "").trim(),
    updateTime: formatUpdateTime(data.updateTime),
  };
}

export async function getDetailedCurrentWeather(): Promise<DetailedCurrentWeather> {
  const data = await hkoJson<Record<string, unknown>>("rhrread");
  const rainfall = asRecord(data.rainfall);

  return {
    updateTime: formatUpdateTime(data.updateTime),
    temperatures: formatStationReadings(data.temperature),
    humidity: formatStationReadings(data.humidity),
    uvIndex: formatUvIndex(data.uvindex),
    rainfallByDistrict: formatRainfallByDistrict(rainfall),
    rainfallPeriod: formatRainfallPeriod(rainfall),
    warningMessage: formatWeatherWarningMessage(data.warningMessage),
    tropicalCycloneMessage: formatTextList(data.tcmessage),
    weatherIconUpdateTime: formatUpdateTime(data.iconUpdateTime),
  };
}

export async function getSpecialWeatherTips(): Promise<SpecialWeatherTips> {
  const data = await hkoJson<Record<string, unknown>>("swt");
  const entries = Array.isArray(data.swt) ? data.swt : [];
  const tips: string[] = [];

  for (const raw of entries) {
    const item = asRecord(raw);
    const desc = String(item?.desc || "").trim();
    if (desc) {
      tips.push(desc);
    }
  }

  const first = asRecord(entries[0]);
  return {
    tips,
    updateTime: formatUpdateTime(first?.updateTime),
  };
}

export async function getNineDayForecastOverview(): Promise<NineDayForecastOverview> {
  const data = await hkoJson<Record<string, unknown>>("fnd");
  const forecasts = Array.isArray(data.weatherForecast) ? data.weatherForecast : [];
  const today = todayInHongKong();
  const dayForecasts = forecasts
    .map((raw) => asRecord(raw))
    .filter((item): item is Record<string, unknown> => item !== undefined)
    .map((item) => formatDayForecast(item));

  const todayForecast = dayForecasts.find((item) => item.date === today) ?? dayForecasts[0] ?? null;
  const upcomingDays = dayForecasts.filter((item) => item.date !== todayForecast?.date).slice(0, 3);

  return {
    generalSituation: String(data.generalSituation || "").trim(),
    seaTemperature: formatSeaTemperature(data.seaTemp),
    soilTemperature: formatSoilTemperature(data.soilTemp),
    updateTime: formatUpdateTime(data.updateTime),
    today: todayForecast,
    upcomingDays,
  };
}

export async function getDetailedDailyWeatherContext(): Promise<DailyDetailedContext> {
  const [current, localForecast, nineDay, specialTips, warnings] = await Promise.all([
    getDetailedCurrentWeather(),
    getLocalForecast(),
    getNineDayForecastOverview(),
    getSpecialWeatherTips(),
    getWarningSnapshot(),
  ]);

  return {
    current,
    localForecast,
    nineDay,
    specialTips,
    warnings,
  };
}

export function buildDetailedDailyBriefing(context: DailyDetailedContext): string {
  const lines: string[] = ["以下為香港天文台詳盡天氣資料：", ""];
  const { current, localForecast, nineDay, specialTips, warnings } = context;

  lines.push(
    "【現時天氣報告】",
    `更新時間：${current.updateTime}`,
    `各區氣溫：${formatStationList(current.temperatures)}`,
    `相對濕度：${formatStationList(current.humidity)}`,
    `紫外線指數：${current.uvIndex}`,
    `過去一小時雨量（${current.rainfallPeriod}）：${formatStationList(current.rainfallByDistrict)}`,
    `天氣圖示更新時間：${current.weatherIconUpdateTime}`,
    `特別天氣提示：${current.warningMessage}`,
  );

  if (current.tropicalCycloneMessage) {
    lines.push(`熱帶氣旋相關消息：${current.tropicalCycloneMessage}`);
  }

  lines.push(
    "",
    "【本港地區天氣預測】",
    `更新時間：${localForecast.updateTime}`,
    `天氣概況：${localForecast.generalSituation || "無"}`,
    `熱帶氣旋資訊：${localForecast.tcInfo || "無"}`,
    `火災危險警告：${localForecast.fireDangerWarning || "無"}`,
    `${localForecast.forecastPeriod || "預測"}：${localForecast.forecastDesc || "無"}`,
    `展望：${localForecast.outlook || "無"}`,
  );

  lines.push(
    "",
    "【九天天氣預報概況】",
    `更新時間：${nineDay.updateTime}`,
    `天氣概況：${nineDay.generalSituation || "無"}`,
    `海水溫度：${nineDay.seaTemperature}`,
    `土壤溫度：${nineDay.soilTemperature}`,
  );

  if (nineDay.today) {
    lines.push("", formatDayForecastBlock("今日預報", nineDay.today));
  }

  for (const day of nineDay.upcomingDays) {
    lines.push("", formatDayForecastBlock("短期預報", day));
  }

  if (specialTips.tips.length > 0) {
    lines.push("", "【特別天氣消息】", `更新時間：${specialTips.updateTime}`);
    for (const tip of specialTips.tips) {
      lines.push(`- ${tip}`);
    }
  }

  if (warnings.hasNotification && warnings.message) {
    lines.push("", "【生效天氣警告】", warnings.message);
  }

  return lines.join("\n");
}

export function buildDetailedCurrentBriefing(context: DailyDetailedContext): string {
  const lines: string[] = ["以下為香港天文台現時及相關天氣資料：", ""];
  const { current, localForecast, nineDay, specialTips, warnings } = context;

  lines.push(
    "【現時天氣報告】",
    `更新時間：${current.updateTime}`,
    `各區氣溫：${formatStationList(current.temperatures)}`,
    `相對濕度：${formatStationList(current.humidity)}`,
    `紫外線指數：${current.uvIndex}`,
    `過去一小時雨量（${current.rainfallPeriod}）：${formatStationList(current.rainfallByDistrict)}`,
    `天氣圖示更新時間：${current.weatherIconUpdateTime}`,
    `特別天氣提示：${current.warningMessage}`,
  );

  if (current.tropicalCycloneMessage) {
    lines.push(`熱帶氣旋相關消息：${current.tropicalCycloneMessage}`);
  }

  lines.push(
    "",
    "【本港地區天氣預測】",
    `更新時間：${localForecast.updateTime}`,
    `天氣概況：${localForecast.generalSituation || "無"}`,
    `熱帶氣旋資訊：${localForecast.tcInfo || "無"}`,
    `火災危險警告：${localForecast.fireDangerWarning || "無"}`,
    `${localForecast.forecastPeriod || "預測"}：${localForecast.forecastDesc || "無"}`,
    `展望：${localForecast.outlook || "無"}`,
  );

  if (nineDay.today) {
    lines.push("", formatDayForecastBlock("今日餘下時間參考", nineDay.today));
  }

  if (specialTips.tips.length > 0) {
    lines.push("", "【特別天氣消息】", `更新時間：${specialTips.updateTime}`);
    for (const tip of specialTips.tips) {
      lines.push(`- ${tip}`);
    }
  }

  if (warnings.hasNotification && warnings.message) {
    lines.push("", "【生效天氣警告】", warnings.message);
  }

  return lines.join("\n");
}

export async function getTodayDayForecast(): Promise<DayForecast | null> {
  const data = await hkoJson<Record<string, unknown>>("fnd");
  const forecasts = Array.isArray(data.weatherForecast) ? data.weatherForecast : [];
  const today = todayInHongKong();

  for (const raw of forecasts) {
    const item = asRecord(raw);
    if (!item || String(item.forecastDate || "") !== today) {
      continue;
    }

    return formatDayForecast(item);
  }

  const first = asRecord(forecasts[0]);
  return first ? formatDayForecast(first) : null;
}

export async function getDailyWeatherContext(): Promise<DailyWeatherContext> {
  const [current, localForecast, todayForecast, warnings] = await Promise.all([
    getCurrentWeather(),
    getLocalForecast(),
    getTodayDayForecast(),
    getWarningSnapshot(),
  ]);

  return {
    current,
    localForecast,
    todayForecast,
    warnings,
  };
}

function formatStationReadings(raw: unknown, valueSuffix = ""): StationReading[] {
  const container = asRecord(raw);
  const records = Array.isArray(container?.data) ? container.data : [];
  const unit = String(container?.unit || "");

  return records
    .map((record) => {
      const item = asRecord(record);
      if (!item || item.value === undefined || item.value === null) {
        return null;
      }

      const place = String(item.place || "未知地區");
      let valueText = String(item.value);
      if (unit === "C") {
        valueText = `${valueText}°C`;
      } else if (unit === "percent") {
        valueText = `${valueText}%`;
      } else if (unit) {
        valueText = `${valueText}${unit}`;
      } else if (valueSuffix) {
        valueText = `${valueText}${valueSuffix}`;
      }

      return { place, value: valueText };
    })
    .filter((item): item is StationReading => item !== null);
}

function formatStationList(readings: StationReading[]): string {
  if (readings.length === 0) {
    return "未能取得";
  }

  return readings.map((reading) => `${reading.place} ${reading.value}`).join("；");
}

function formatRainfallByDistrict(container: Record<string, unknown> | undefined): StationReading[] {
  const records = Array.isArray(container?.data) ? container.data.filter(isRecord) : [];
  const unit = String(container?.unit || "mm");

  return records.map((record) => ({
    place: String(record.place || "未知地區"),
    value: `${String(record.max ?? 0)}${unit}`,
  }));
}

function formatRainfallPeriod(container: Record<string, unknown> | undefined): string {
  const start = container?.startTime ? formatDateTime(String(container.startTime)) : "";
  const end = container?.endTime ? formatDateTime(String(container.endTime)) : "";
  if (start && end) {
    return `${start} 至 ${end}`;
  }
  return "未知";
}

function formatSeaTemperature(raw: unknown): string {
  const record = asRecord(raw);
  if (!record || record.value === undefined || record.value === null) {
    return "未能取得";
  }

  const place = String(record.place || "未知地點");
  const unit = String(record.unit || "C");
  const time = record.recordTime ? formatDateTime(String(record.recordTime)) : "";
  const value = `${place} ${String(record.value)}°${unit}`;
  return time ? `${value}（${time}）` : value;
}

function formatSoilTemperature(raw: unknown): string {
  if (!Array.isArray(raw) || raw.length === 0) {
    return "未能取得";
  }

  return raw
    .map((item) => {
      const record = asRecord(item);
      if (!record || record.value === undefined || record.value === null) {
        return null;
      }

      const place = String(record.place || "未知地點");
      const depth = asRecord(record.depth);
      const depthValue = depth?.value !== undefined ? `${String(depth.value)}${String(depth.unit || "m")}` : "";
      const unit = String(record.unit || "C");
      const depthLabel = depthValue ? ` 深度${depthValue}` : "";
      return `${place}${depthLabel} ${String(record.value)}°${unit}`;
    })
    .filter(Boolean)
    .join("；");
}

function formatTextList(raw: unknown): string {
  if (!Array.isArray(raw) || raw.length === 0) {
    return "";
  }

  return raw
    .map((item) => String(item).trim())
    .filter(Boolean)
    .join("\n");
}

function formatDayForecastBlock(label: string, day: DayForecast): string {
  return [
    `【${label}：${day.date}（${day.week}）】`,
    `天氣：${day.weather}`,
    `氣溫：${day.minTemp} - ${day.maxTemp}`,
    `相對濕度：${day.minRh} - ${day.maxRh}`,
    `風向風速：${day.wind}`,
    `降雨概率：${day.rainProbability || "無"}`,
  ].join("\n");
}

function formatDayForecast(item: Record<string, unknown>): DayForecast {
  const maxTemp = asRecord(item.forecastMaxtemp);
  const minTemp = asRecord(item.forecastMintemp);
  const maxRh = asRecord(item.forecastMaxrh);
  const minRh = asRecord(item.forecastMinrh);

  return {
    date: String(item.forecastDate || ""),
    week: String(item.week || ""),
    weather: String(item.forecastWeather || "").trim(),
    minTemp: formatTempValue(minTemp),
    maxTemp: formatTempValue(maxTemp),
    minRh: formatPercentValue(minRh),
    maxRh: formatPercentValue(maxRh),
    wind: String(item.forecastWind || "").trim(),
    rainProbability: String(item.PSR || "").trim(),
  };
}

function formatTempValue(record: Record<string, unknown> | undefined): string {
  if (!record || record.value === undefined || record.value === null) {
    return "未能取得";
  }

  return `${String(record.value)}°${String(record.unit || "C")}`;
}

function formatPercentValue(record: Record<string, unknown> | undefined): string {
  if (!record || record.value === undefined || record.value === null) {
    return "未能取得";
  }

  const unit = String(record.unit || "percent");
  return unit === "percent" ? `${String(record.value)}%` : `${String(record.value)}${unit}`;
}

function todayInHongKong(): string {
  const formatter = new Intl.DateTimeFormat("en-CA", {
    timeZone: "Asia/Hong_Kong",
    year: "numeric",
    month: "2-digit",
    day: "2-digit",
  });

  return formatter.format(new Date()).replace(/-/g, "");
}

async function hkoJson<T>(dataType: string): Promise<T> {
  const url = new URL(HKO_BASE_URL);
  url.searchParams.set("dataType", dataType);
  url.searchParams.set("lang", "tc");

  const response = await fetch(url.toString());
  if (!response.ok) {
    throw new Error(`HKO API ${dataType} failed with ${response.status}`);
  }

  return response.json() as Promise<T>;
}

function formatTemperature(raw: unknown): string {
  const container = asRecord(raw);
  const records = Array.isArray(container?.data) ? container.data : [];
  const hkoRecord = records.map(asRecord).find((record) => record?.place === "香港天文台");
  const record = hkoRecord ?? asRecord(records[0]);

  if (!record || record.value === undefined || record.value === null) {
    return "未能取得";
  }

  const place = String(record.place || "香港");
  const unit = String(container?.unit || "C");
  return `${place} ${String(record.value)}°${unit}`;
}

function formatHumidity(raw: unknown): string {
  const container = asRecord(raw);
  const records = Array.isArray(container?.data) ? container.data : [];
  const record = asRecord(records[0]);
  if (!record || record.value === undefined || record.value === null) {
    return "未能取得";
  }

  return `${String(record.value)}${String(container?.unit || "%")}`;
}

function formatUvIndex(raw: unknown): string {
  const container = asRecord(raw);
  const records = Array.isArray(container?.data) ? container.data : [];
  const record = asRecord(records[0]);
  if (!record || record.value === undefined || record.value === null) {
    return "未能取得";
  }

  return record.desc ? `${String(record.value)}（${String(record.desc)}）` : String(record.value);
}

function formatRainfall(raw: unknown): string {
  const container = asRecord(raw);
  const records = Array.isArray(container?.data) ? container.data.filter(isRecord) : [];
  const rainyRecords = records.filter((record) => {
    const maximum = record.max;
    return maximum !== 0 && maximum !== "0" && maximum !== undefined && maximum !== null;
  });

  if (rainyRecords.length === 0) {
    return "過去一小時大部分地區沒有錄得雨量";
  }

  const unit = String(container?.unit || "mm");
  return rainyRecords
    .slice(0, 5)
    .map((record) => {
      const place = String(record.place || "未知地區");
      if (record.min === record.max || record.min === undefined || record.min === "") {
        return `${place} ${String(record.max)}${unit}`;
      }
      return `${place} ${String(record.min)}-${String(record.max)}${unit}`;
    })
    .join("；");
}

function formatWeatherWarningMessage(raw: unknown): string {
  if (!Array.isArray(raw) || raw.length === 0) {
    return "沒有特別天氣提示";
  }

  const text = raw
    .map((item) => String(item).trim())
    .filter(Boolean)
    .join("\n");
  return text || "沒有特別天氣提示";
}

function formatUpdateTime(raw: unknown): string {
  return raw ? formatDateTime(String(raw)) : "未知";
}

function formatDateTime(raw: string): string {
  const date = new Date(raw);
  if (Number.isNaN(date.getTime())) {
    return raw;
  }

  const formatter = new Intl.DateTimeFormat("sv-SE", {
    timeZone: "Asia/Hong_Kong",
    year: "numeric",
    month: "2-digit",
    day: "2-digit",
    hour: "2-digit",
    minute: "2-digit",
    hour12: false,
  });
  return formatter.format(date);
}

function asRecord(value: unknown): Record<string, unknown> | undefined {
  return isRecord(value) ? value : undefined;
}

function isRecord(value: unknown): value is Record<string, unknown> {
  return typeof value === "object" && value !== null && !Array.isArray(value);
}
